proposals/NNNN-package-registry-auth.md

```diff
@@ -0,0 +1,387 @@
+`.netrc` file. (Only macOS Keychain will be supported in the 
+initial feature release; more might be added in the future.)
+
+See [credential storage](#credential-storage) for more details on configuring 
+credentials for each authentication type.
+
+### Credential storage
+
+#### Basic Authentication
+
+##### macOS Keychain
+
+Registry credentials should be stored as "Internet password" 
+items in the macOS Keychain. The "item name" should be the 
+registry URL, including `https://` (e.g., `https://example-registry.com`).
+
+##### `.netrc` file (non-macOS platforms only)
+
+A `.netrc` entry for basic authentication looks as follows:
+
+```
+machine example-registry.com
+login jappleseed
+password alpine
+```
+
+By default, SwiftPM looks for `.netrc` file in the user's 
+home directory. A custom `.netrc` file can be specified using 
+the `--netrc-file` option.
+
+#### Token Authentication
+
+User can configure access token for a registry as similarly 
+done for basic authentication, but with `token` as the login/username 
+and the access token as the password. 
+
+For example, a `.netrc` entry would look like:
+
+```
+machine example-registry.com
+login token
+password jappleseedstoken
+```
+
+### Additional changes in SwiftPM
+
+1. Only the user-level `.netrc` file will be used. Project-level `.netrc` file will not be supported.
+2. SwiftPM will perform lookups in one credential store only. For macOS, it will be Keychain. For all other platforms, it will be the user-level `.netrc` file.
+3. The `--disable-keychain` and `--disable-netrc` options will be removed.
+
+### New package registry service API
+
+A package registry that requires authentication must implement
+the new API endpoint(s) covered in this section.
+
+#### `login` API
+
+SwiftPM will send a HTTP `POST` request to `/login` to validate 
+user credentials provided by the `login` subcommand. The request 
+will include an `Authorization` HTTP header constructed as follows:
+
+* Basic authentication: `Authorization: Basic <base64 encoded username:password>`
+* Token authentication: `Authorization: Bearer <token>`
+
+The registry service must return HTTP status `200` in the 
+response if login is successful, and `401` otherwise.
+
+In case the registry service does not support an authentication method,
+it should return HTTP status `501`.
+
+SwiftPM will persist user credentials to local credential store 
+if login is successful.
+
+## Security
+
+This proposal moves SwiftPM to use operating system's native credential 
+store (e.g., macOS Keychain) on supported platforms, which should yield
+better security.
+
+We are also eliminating the use of project-level `.netrc` file. This should
+prevent accidental checkin of `.netrc` file and thus leakage of sensitive
+information.
+
+## Impact on existing packages
+
+This proposal eliminates the project-level `.netrc` file. There should be 
+no other impact on existing packages.
+
```

Comment:
Is there another impact for users that rely on netrc today on macOS that must migrate to keychain ?